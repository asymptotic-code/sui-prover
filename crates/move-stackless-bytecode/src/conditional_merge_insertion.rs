// Copyright (c) Asymptotic
// SPDX-License-Identifier: Apache-2.0

//! Inserts `if_then_else` at merge points for simple consecutive if-statements
//! while keeping original control flow intact. This is intended to make merge
//! semantics explicit for later verification/optimization passes.
//!
//! Supported pattern per if-block:
//! - Bytecode::Branch(then_label, else_label, cond)
//! - then_label block ends with Jump to else_label
//! - else_label is the merge label (fallthrough path)
//! - Simplifying assumption: pick the last Assign in the then block as
//!   the then-value producer for the merged variable, and the last assignment
//!   to the same variable before the Branch as the else-value.
//!
//! The pass inserts at the merge label (right after the Label):
//!   t_new := if_then_else(cond, then_src, else_src)
//!   var   := t_new
//!
//! This leaves branch-local assignments in place (they may be dead after merge
//! and can be cleaned up by later passes).

use crate::{
    function_target::FunctionData,
    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
    graph::{DomRelation, Graph},
    stackless_bytecode::{AttrId, Bytecode, Operation},
    stackless_control_flow_graph::{BlockContent, BlockId, StacklessControlFlowGraph},
};
use move_model::model::FunctionEnv;

pub struct ConditionalMergeInsertionProcessor {
    debug: bool,
}

impl ConditionalMergeInsertionProcessor {
    pub fn new() -> Box<Self> {
        Box::new(Self { debug: false })
    }

    #[allow(dead_code)]
    pub fn new_with_debug() -> Box<Self> {
        Box::new(Self { debug: true })
    }

    fn next_attr_pair(data: &FunctionData) -> (AttrId, AttrId) {
        let base = data.next_free_attr_index();
        (AttrId::new(base), AttrId::new(base + 1))
    }

    // Dominator-based merge detection for a branch at pc.
    // Returns (cond_temp, then_start_pc, merge_block_start_pc) if we can determine a merge.
    //
    // Recall the immediate postdominator of the branch corresponds to the
    // structural join. For SSA-style placement of `if_then_else` (phi) per
    // variable, we compute the dominance frontier of each definition and place
    // merges at those DF nodes. This function currently finds the structural
    // join (later phases can refine per-var merge placement using DF).
    fn find_merge_by_dominators(
        code: &[Bytecode],
        back_cfg: &StacklessControlFlowGraph,
        branch_pc: usize,
    ) -> Option<(usize, usize, usize)> {
        let (then_label, cond) = match &code[branch_pc] {
            Bytecode::Branch(_, tl, _el, cond) => (tl, *cond as usize),
            _ => return None,
        };

        let labels = Bytecode::label_offsets(code);
        let then_pc = *labels.get(then_label)? as usize;
        let branch_block = Self::pc_to_block(back_cfg, branch_pc as u16)?;

        // Compute merge block as immediate post-dominator of the branch block using
        // reversed-graph dominator analysis (postdominators of the original graph).
        let merge_block = Self::find_immediate_post_dominator(back_cfg, branch_block)?;
        let merge_pc = Self::block_start_pc(back_cfg, merge_block)? as usize;

        Some((cond, then_pc, merge_pc))
    }

    fn block_start_pc(cfg: &StacklessControlFlowGraph, block: BlockId) -> Option<u16> {
        match cfg.content(block) {
            BlockContent::Basic { lower, .. } => Some(*lower),
            BlockContent::Dummy => None,
        }
    }

    fn pc_to_block(cfg: &StacklessControlFlowGraph, pc: u16) -> Option<BlockId> {
        for b in cfg.blocks() {
            match cfg.content(b) {
                BlockContent::Basic { lower, upper } => {
                    if *lower <= pc && pc <= *upper {
                        return Some(b);
                    }
                }
                BlockContent::Dummy => {}
            }
        }
        None
    }

    fn find_immediate_post_dominator(
        back_cfg: &StacklessControlFlowGraph,
        branch_block: BlockId,
    ) -> Option<BlockId> {
        // Build reversed graph and compute dominators (postdominators of original CFG)
        let entry = back_cfg.entry_block();
        let nodes = back_cfg.blocks();
        let edges: Vec<(BlockId, BlockId)> = nodes
            .iter()
            .flat_map(|x| {
                back_cfg
                    .successors(*x)
                    .iter()
                    .map(|y| (*x, *y))
                    .collect::<Vec<_>>()
            })
            .collect();
        let graph = Graph::new(entry, nodes.clone(), edges);
        let dom_rev = DomRelation::new(&graph);

        // Candidates are postdominators of branch_block (including itself and exit)
        let candidates: Vec<BlockId> = nodes
            .into_iter()
            .filter(|b| {
                *b != branch_block
                    && dom_rev.is_reachable(*b)
                    && dom_rev.is_dominated_by(branch_block, *b)
            })
            .collect();

        if candidates.is_empty() {
            return None;
        }

        // Immediate postdominator is the next node required on any path to exit--
        // i.e., deepest candidate: it should not dominate any other candidate.
        for &c in &candidates {
            let mut dominates_any = false;
            for &o in &candidates {
                if o != c && dom_rev.is_dominated_by(o, c) {
                    // c dominates o in reversed graph => c is above; skip it
                    dominates_any = true;
                    break;
                }
            }
            if !dominates_any {
                return Some(c);
            }
        }
        // Fallback
        candidates.into_iter().next()
    }

    // Return the last assignment (dst:=src) in the then block before merge
    fn then_branch_last_var_assign(
        code: &[Bytecode],
        then_start_pc: usize,
        merge_label_pc: usize,
    ) -> Option<(usize /*var*/, usize /*then_src*/)> {
        let mut result: Option<(usize, usize)> = None;
        let mut pc = then_start_pc;
        while pc < code.len() && pc < merge_label_pc {
            match &code[pc] {
                Bytecode::Assign(_, dst, src, _) => {
                    result = Some((*dst as usize, *src as usize));
                }
                // If we see a call writing directly to a var, take var itself as then_src
                Bytecode::Call(_, dests, _op, _srcs, _) if !dests.is_empty() => {
                    let var = dests[0] as usize;
                    result = Some((var, var));
                }
                _ => {}
            }
            if code[pc].is_unconditional_branch() {
                break;
            }
            pc += 1;
        }
        result
    }

    // Scan backwards before `branch_pc` to find the last assignment to `var` for else-path.
    fn else_var_before_branch(code: &[Bytecode], branch_pc: usize, var: usize) -> Option<usize> {
        let mut pc: isize = branch_pc as isize - 1;
        while pc >= 0 {
            match &code[pc as usize] {
                Bytecode::Assign(_, dst, src, _) if *dst as usize == var => {
                    return Some(*src as usize);
                }
                _ => {}
            }
            pc -= 1;
        }
        None
    }
}

impl FunctionTargetProcessor for ConditionalMergeInsertionProcessor {
    fn process(
        &self,
        _targets: &mut FunctionTargetsHolder,
        _func_env: &FunctionEnv,
        mut data: FunctionData,
        _scc_opt: Option<&[FunctionEnv]>,
    ) -> FunctionData {
        let mut pc = 0;
        while pc < data.code.len() {
            if matches!(data.code[pc], Bytecode::Branch(..)) {
                // Build backward CFG after any mutations to code to keep mapping fresh
                let back_cfg = StacklessControlFlowGraph::new_backward(&data.code, false);
                if let Some((cond, then_start, merge_pc)) =
                    Self::find_merge_by_dominators(&data.code, &back_cfg, pc)
                {
                    if let Some((var, then_src)) =
                        Self::then_branch_last_var_assign(&data.code, then_start, merge_pc)
                    {
                        if let Some(else_src) = Self::else_var_before_branch(&data.code, pc, var) {
                            let insert_at =
                                if matches!(data.code.get(merge_pc), Some(Bytecode::Label(..))) {
                                    // In case of label / jump dest, insert after.
                                    merge_pc + 1
                                } else {
                                    merge_pc
                                };

                            // Allocate a fresh temp of same type as `var`
                            let new_temp = data.local_types.len();
                            let var_ty = data.local_types[var].clone();
                            data.local_types.push(var_ty);

                            let (ite_attr, assign_attr) = Self::next_attr_pair(&data);

                            // t_new := if_then_else(cond, then_src, else_src)
                            let ite = Bytecode::Call(
                                ite_attr,
                                vec![new_temp],
                                Operation::IfThenElse,
                                vec![cond, then_src, else_src],
                                None,
                            );
                            // var := t_new
                            let assign = Bytecode::Assign(
                                assign_attr,
                                var,
                                new_temp,
                                crate::stackless_bytecode::AssignKind::Copy,
                            );

                            if self.debug {
                                eprintln!(
                                    "[conditional_merge_insertion] insert@{}: t{} := if_then_else(t{}, t{}, t{}); t{} := t{}",
                                    insert_at, new_temp, cond, then_src, else_src, var, new_temp
                                );
                            }

                            data.debug_comments.insert(
                                ite_attr,
                                format!(
                                    "conditional_merge_insertion: t{} := if_then_else(t{}, t{}, t{})",
                                    new_temp, cond, then_src, else_src
                                ),
                            );
                            data.debug_comments.insert(
                                assign_attr,
                                format!(
                                    "conditional_merge_insertion: merge assign t{} := t{}",
                                    var, new_temp
                                ),
                            );

                            // Insert and skip past assign
                            data.code.splice(insert_at..insert_at, [ite, assign]);
                            pc = insert_at + 2;
                            continue;
                        }
                    }
                }
            }

            pc += 1;
        }

        data
    }

    fn name(&self) -> String {
        "conditional_merge_insertion".to_string()
    }
}
