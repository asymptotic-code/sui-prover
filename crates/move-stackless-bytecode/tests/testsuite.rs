// Copyright (c) The Diem Core Contributors
// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

use anyhow::anyhow;
use codespan_reporting::{diagnostic::Severity, term::termcolor::Buffer};
use move_command_line_common::insta_assert;
use move_compiler::{diagnostics::warning_filters::WarningFiltersBuilder, shared::PackagePaths};
use move_model::{model::GlobalEnv, options::ModelBuilderOptions, run_model_builder_with_options};
use move_stackless_bytecode::{
    borrow_analysis::BorrowAnalysisProcessor,
    clean_and_optimize::CleanAndOptimizeProcessor,
    eliminate_imm_refs::EliminateImmRefsProcessor,
    escape_analysis::EscapeAnalysisProcessor,
    function_target_pipeline::{
        FunctionTargetPipeline, FunctionTargetsHolder, ProcessorResultDisplay,
    },
    livevar_analysis::LiveVarAnalysisProcessor,
    memory_instrumentation::MemoryInstrumentationProcessor,
    mut_ref_instrumentation::MutRefInstrumenter,
    options::ProverOptions,
    print_targets_for_test,
    reaching_def_analysis::ReachingDefProcessor,
};
use regex::Regex;
use std::{fs::File, io::Read, path::Path};

// Extracts lines out of some text file where each line starts with `start` which can be a regular
// expressions. Returns the list of such lines with `start` stripped. Use as in
// `extract_test_directives(file, "// dep:")`.
fn extract_test_directives(path: &Path, start: &str) -> anyhow::Result<Vec<String>> {
    let rex = Regex::new(&format!("(?m)^{}(?P<ann>.*?)$", start)).unwrap();
    let mut content = String::new();
    let mut file = File::open(path)?;
    file.read_to_string(&mut content)?;
    let mut at = 0;
    let mut res = vec![];
    while let Some(cap) = rex.captures(&content[at..]) {
        res.push(cap.name("ann").unwrap().as_str().trim().to_string());
        at += cap.get(0).unwrap().end();
    }
    Ok(res)
}

fn get_tested_transformation_pipeline(
    dir_name: &str,
) -> anyhow::Result<Option<FunctionTargetPipeline>> {
    match dir_name {
        "from_move" => Ok(None),
        "eliminate_imm_refs" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            Ok(Some(pipeline))
        }
        "mut_ref_instrumentation" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            Ok(Some(pipeline))
        }
        "reaching_def" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            Ok(Some(pipeline))
        }
        "conditional_merge_insertion" => {
            use move_stackless_bytecode::conditional_merge_insertion::ConditionalMergeInsertionProcessor;
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(ConditionalMergeInsertionProcessor::new_with_debug());
            Ok(Some(pipeline))
        }
        "livevar" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "borrow" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "borrow_strong" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            Ok(Some(pipeline))
        }
        "escape_analysis" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(Box::new(EscapeAnalysisProcessor {}));
            Ok(Some(pipeline))
        }
        "memory_instr" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            pipeline.add_processor(MemoryInstrumentationProcessor::new());
            Ok(Some(pipeline))
        }
        "clean_and_optimize" => {
            let mut pipeline = FunctionTargetPipeline::default();
            pipeline.add_processor(EliminateImmRefsProcessor::new());
            pipeline.add_processor(MutRefInstrumenter::new());
            pipeline.add_processor(ReachingDefProcessor::new());
            pipeline.add_processor(LiveVarAnalysisProcessor::new());
            pipeline.add_processor(BorrowAnalysisProcessor::new());
            pipeline.add_processor(MemoryInstrumentationProcessor::new());
            pipeline.add_processor(CleanAndOptimizeProcessor::new());
            Ok(Some(pipeline))
        }
        _ => Err(anyhow!(
            "the sub-directory `{}` has no associated pipeline to test",
            dir_name
        )),
    }
}

fn test_runner(path: &Path) -> datatest_stable::Result<()> {
    // Allow opting out of external deps (e.g., move-stdlib) so isolated tests can run
    // without requiring a local stdlib checkout. Set env var `STACKLESS_TEST_IGNORE_DEPS=1`.
    let ignore_deps = std::env::var("STACKLESS_TEST_IGNORE_DEPS").is_ok();
    let mut sources = if ignore_deps {
        vec![path.to_string_lossy().to_string()]
    } else {
        let mut deps = extract_test_directives(path, "// dep:")?;
        // Keep only deps that exist on disk to avoid hard failures on missing stdlib
        deps.retain(|p| std::path::Path::new(p).exists());
        deps.push(path.to_string_lossy().to_string());
        deps
    };
    let env: GlobalEnv = run_model_builder_with_options(
        vec![PackagePaths {
            name: None,
            paths: sources,
            named_address_map: move_stdlib::named_addresses(),
        }],
        vec![],
        ModelBuilderOptions::default(),
        Some(WarningFiltersBuilder::unused_warnings_filter_for_test()),
    )?;
    let out = if env.has_errors() {
        let mut error_writer = Buffer::no_color();
        env.report_diag(&mut error_writer, Severity::Error);
        String::from_utf8_lossy(&error_writer.into_inner()).to_string()
    } else {
        let options = ProverOptions {
            stable_test_output: true,
            ..Default::default()
        };
        env.set_extension(options);
        let dir_name = path
            .parent()
            .and_then(|p| p.file_name())
            .and_then(|p| p.to_str())
            .ok_or_else(|| anyhow!("bad file name"))?;
        let pipeline_opt = get_tested_transformation_pipeline(dir_name)?;

        // Initialize and print function targets
        let mut text = String::new();
        let mut targets = FunctionTargetsHolder::new(None);
        for module_env in env.get_modules() {
            for func_env in module_env.get_functions() {
                targets.add_target(&func_env);
            }
        }
        let show_livevars = std::env::var("STACKLESS_TEST_SHOW_LIVENESS").is_ok();
        let show_borrow = true;
        let show_reach = std::env::var("STACKLESS_TEST_SHOW_REACHING_DEFS").is_ok();
        text += &move_stackless_bytecode::print_targets_for_test_with_flags(
            &env,
            "initial translation from Move",
            &targets,
            show_livevars,
            show_borrow,
            show_reach,
        );

        // Run pipeline if any
        if let Some(pipeline) = pipeline_opt {
            let _ = pipeline.run(&env, &mut targets);
            let processor = pipeline.last_processor();
            if !processor.is_single_run() {
                text += &move_stackless_bytecode::print_targets_for_test_with_flags(
                    &env,
                    &format!("after pipeline `{}`", dir_name),
                    &targets,
                    show_livevars,
                    show_borrow,
                    show_reach,
                );
            }
            text += &ProcessorResultDisplay {
                env: &env,
                targets: &targets,
                processor,
            }
            .to_string();
        }
        // add Warning and Error diagnostics to output
        let mut error_writer = Buffer::no_color();
        if env.has_errors() || env.has_warnings() {
            env.report_diag(&mut error_writer, Severity::Warning);
            text += "============ Diagnostics ================\n";
            text += &String::from_utf8_lossy(&error_writer.into_inner());
        }
        text
    };
    insta_assert! {
        input_path: path,
        contents: out,
    };
    Ok(())
}

datatest_stable::harness! {
    { test = test_runner, root = "tests", pattern = r".*\.move" },
}
