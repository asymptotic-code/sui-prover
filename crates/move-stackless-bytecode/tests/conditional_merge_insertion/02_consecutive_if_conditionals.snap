---
source: crates/move-stackless-bytecode/tests/testsuite.rs
---
============ initial translation from Move ================

[variant baseline]
public fun test::f($t0|a: u64): u64 {
     var $t1|result#1#0: u64
     var $t2: u64
     var $t3: u64
     var $t4: u64
     var $t5: bool
     var $t6: u64
     var $t7: u64
     var $t8: u64
     var $t9: u64
     var $t10: u64
     var $t11: bool
     var $t12: u64
     var $t13: u64
     var $t14: u64
     var $t15: u64
     var $t16: u64
     var $t17: bool
     var $t18: u64
     var $t19: u64
     var $t20: u64
     var $t21: u64
  0: $t2 := 0
  1: $t1 := $t2
  2: $t3 := copy($t0)
  3: $t4 := 0
  4: $t5 := >($t3, $t4)
  5: if ($t5) goto 6 else goto 12
  6: label L1
  7: $t6 := move($t1)
  8: $t7 := 10
  9: $t8 := *($t6, $t7)
 10: $t1 := $t8
 11: goto 12
 12: label L0
 13: $t9 := copy($t0)
 14: $t10 := 5
 15: $t11 := >($t9, $t10)
 16: if ($t11) goto 17 else goto 23
 17: label L3
 18: $t12 := move($t1)
 19: $t13 := 20
 20: $t14 := *($t12, $t13)
 21: $t1 := $t14
 22: goto 23
 23: label L2
 24: $t15 := move($t0)
 25: $t16 := 10
 26: $t17 := >($t15, $t16)
 27: if ($t17) goto 28 else goto 34
 28: label L5
 29: $t18 := move($t1)
 30: $t19 := 30
 31: $t20 := *($t18, $t19)
 32: $t1 := $t20
 33: goto 34
 34: label L4
 35: $t21 := move($t1)
 36: return $t21
}

============ after pipeline `conditional_merge_insertion` ================

[variant baseline]
public fun test::f($t0|a: u64): u64 {
     var $t1|result#1#0: u64
     var $t2: u64
     var $t3: u64
     var $t4: bool
     var $t5: u64
     var $t6: u64
     var $t7: bool
     var $t8: u64
     var $t9: u64
     var $t10: bool
     var $t11: u64
     var $t12: u64
     var $t13: u64
     var $t14: u64
  0: $t2 := 0
  1: $t1 := $t2
  2: $t3 := 0
  3: $t4 := >($t0, $t3)
  4: if ($t4) goto 5 else goto 8
  5: label L1
  6: $t5 := 10
  7: $t1 := *($t2, $t5)
  8: label L0
     # conditional_merge_insertion: t12 := if_then_else(t4, t1, t2)
  9: $t12 := if_then_else($t4, $t1, $t2)
     # conditional_merge_insertion: merge assign t1 := t12
 10: $t1 := copy($t12)
 11: $t6 := 5
 12: $t7 := >($t0, $t6)
 13: if ($t7) goto 14 else goto 17
 14: label L3
 15: $t8 := 20
 16: $t1 := *($t1, $t8)
 17: label L2
     # conditional_merge_insertion: t13 := if_then_else(t7, t1, t12)
 18: $t13 := if_then_else($t7, $t1, $t12)
     # conditional_merge_insertion: merge assign t1 := t13
 19: $t1 := copy($t13)
 20: $t9 := 10
 21: $t10 := >($t0, $t9)
 22: if ($t10) goto 23 else goto 26
 23: label L5
 24: $t11 := 30
 25: $t1 := *($t1, $t11)
 26: label L4
     # conditional_merge_insertion: t14 := if_then_else(t10, t1, t13)
 27: $t14 := if_then_else($t10, $t1, $t13)
     # conditional_merge_insertion: merge assign t1 := t14
 28: $t1 := copy($t14)
 29: return $t1
}
